pallet/deposit/src/lib.rs

```diff
@@ -216,14 +217,19 @@ pub mod pallet {
 			}
 
 			if to_claim.0 != 0 {
-				T::Ring::transfer(&account_id(), who, to_claim.0, AllowDeath)?;
+				if let Err(e) = T::Ring::transfer(&T::Treasury::get(), who, to_claim.0, AllowDeath)
+				{
+					Self::deposit_event(Event::MigrationFailedOnRefund);
+
+					Err(e)?;
+				}
+
 				Self::deposit_event(Event::DepositsClaimed {
 					owner: who.clone(),
 					deposits: to_claim.1,
 				});
 			}
 			if to_migrate.0 != 0 {
-				T::Ring::transfer(&account_id(), &T::Treasury::get(), to_migrate.0, AllowDeath)?;
 				T::DepositMigrator::migrate(who.clone(), to_migrate.0, to_migrate.2)?;
```

Comment:
Should not throw errors here right?

Reply:
In FP, we treat any I/O operation as the side effect.

And here, especially we interact with the DB, `take` and `transfer` have already introduced the side effects. We should keep the error and utilize it to help rollback the state.

Reply:
You didn't get what I mean.

If keep the error, this batch (10 items) will be inserted again, but some of them might have been migrated, then they will be keep migrated again and again.

Reply:
Considering that you are on vacation and the likelihood of errors is very low, I suggest we keep things simple. At the same time, we must absolutely avoid any ledger inaccuracies or double-spending in case an error does occur.

> Then we can remove the event and insert the failure items into a different storage, and process that storage manually later. Using storage is easier to track and process compare to event.

If this looks simpler for you, It works for me. (Just need me take another several minutes to review the difference between them, good so far.)